File: HackerRank-in-a-String-.cpp
#include <bits/stdc++.h>

using namespace std;

string ltrim(const string &);
string rtrim(const string &);

/*
 * Complete the 'hackerrankInString' function below.
 *
 * The function is expected to return a STRING.
 * The function accepts STRING s as parameter.
 */

string hackerrankInString(string s)
{
    constexpr std::string_view target = "hackerrank";
    int targetIndex = 0;

    for (char c : s)
    {
        if (targetIndex < target.length() && c == target[targetIndex])
        {
            targetIndex++;
            if (targetIndex == target.length())
            {
                return "YES";
            }
        }
    }

    return "NO";
}


string hackerrankInString2(string s)
{
    std::deque<char> dq = {'h', 'a', 'c', 'k', 'e', 'r', 'r', 'a', 'n', 'k'};
    
    for (const auto & c : s)
    {
        if (c == dq[0])
        {
            dq.pop_front();
            if (dq.empty())
            {
                return "YES";
            }
        }
    }

    return "NO";
}


int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string q_temp;
    getline(cin, q_temp);

    int q = stoi(ltrim(rtrim(q_temp)));

    for (int q_itr = 0; q_itr < q; q_itr++) {
        string s;
        getline(cin, s);

        string result = hackerrankInString(s);

        fout << result << "\n";
    }

    fout.close();

    return 0;
}

File: HackerRank-in-a-String-.cs
using System.CodeDom.Compiler;
using System.Collections.Generic;
using System.Collections;
using System.ComponentModel;
using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.Serialization;
using System.Text.RegularExpressions;
using System.Text;
using System;

class Result
{
    /*
     * Complete the 'hackerrankInString' function below.
     *
     * The function is expected to return a STRING.
     * The function accepts STRING s as parameter.
     */

    public static string hackerrankInString(string s)
    {
        ReadOnlySpan<char> target = "hackerrank";
        int targetIndex = 0;

        foreach (char c in s)
        {
            if (targetIndex < target.Length && c == target[targetIndex])
            {
                targetIndex++;
                if (targetIndex == target.Length)
                {
                    return "YES";
                }
            }
        }

        return "NO";
    }


    public static string hackerrankInString2(string s)
    {
        Queue<char> queue = new Queue<char>("hackerrank");

        foreach (char c in s)
        {
            if (queue.Count > 0 && c == queue.Peek()) 
            {
                queue.Dequeue();
                if (queue.Count == 0)
                {
                    return "YES";
                }
            }
        }

        return "NO";
    }
}


class Solution
{
    public static void Main(string[] args)
    {
        TextWriter textWriter = new StreamWriter(@System.Environment.GetEnvironmentVariable("OUTPUT_PATH"), true);

        int q = Convert.ToInt32(Console.ReadLine().Trim());

        for (int qItr = 0; qItr < q; qItr++)
        {
            string s = Console.ReadLine();

            string result = Result.hackerrankInString(s);

            textWriter.WriteLine(result);
        }

        textWriter.Flush();
        textWriter.Close();
    }
}
